File: cpp/solutions/solution_3.cpp
#include "primes.hpp"
#include "registry.hpp"

int64_t solution_3() {
    int64_t const number = 600851475143L;
    int64_t remainder = number;
    for (auto prime : PrimeGenerator()) {
        while (remainder % prime == 0) {
            remainder /= prime;
        }
        if (remainder == 1) {
            return prime;
        }
    }
    throw std::runtime_error("Unreachable.");
}

static Registration registration(3, solution_3);

File: python/src/project_euler_python/solutions/solution_3.py
from ..primes import PrimeList


number = 600851475143


def solution_reducing() -> int:
    remainder = number
    last_factor = 0
    for prime in PrimeList():
        while remainder % prime == 0:
            last_factor = prime
            remainder /= prime
        if remainder == 1:
            break
    return last_factor
